File: java/verify_preorder_serialization_of_a_binary_tree.java
public class Solution {
    public boolean isValidSerialization(String preorder) {
        List<String> stack=new ArrayList<String>();
        String[] nodes = preorder.split(",");
        
        for (int i=0; i<nodes.length; i++) {
            stack.add(nodes[i]);
            while (stack.size()>=3) {
                if (stack.get(stack.size()-1).equals("#") && 
                    stack.get(stack.size()-2).equals("#")) {
                    if (stack.get(stack.size()-3).equals("#")) return false;
                    else {
                        stack.subList(stack.size()-3, stack.size()).clear();
                        stack.add("#");
                    }
                } else {
                    break;
                }
            }
        }
        
        if (stack.size()==1 && stack.get(0).equals("#")) return true;
        else return false;
    }
}

File: python/verify-preorder-serialization-of-a-binary-tree.py
class Solution:
    def isValidSerialization(self, preorder):
        """
        :type preorder: str
        :rtype: bool
        """
        stk=[]
        preorder=preorder.split(',')
        for p in preorder:
            if p=="#":
                while len(stk)>0 and stk[-1]=="#":
                    stk.pop()
                    if len(stk)==0: return False
                    stk.pop()
            stk.append(p)
            
        if len(stk)==1 and stk[-1]=="#": return True
        else: return False
